packages/react/src/hooks/useTabs.ts

```diff
@@ -0,0 +1,229 @@
+import {useCallback, useState, useRef, type HTMLAttributes, type KeyboardEvent} from 'react'
+import {useId} from '@reach/auto-id'
+
+type TabOrientation = 'horizontal' | 'vertical'
+
+type UseTabsOptions = {
+  defaultTab?: string
+  autoActivate?: boolean
+  onTabActivate?: (id: string) => void
+  orientation?: TabOrientation
+}
+
+type TabState = {
+  activeTab: string | null
+  focusedTab: string | null
+  tabs: Set<string>
+}
+
+type TabListProps = {
+  label?: string
+  labelledBy?: string
+}
+
+type UseTabs = {
+  activeTab: string | null
+  focusedTab: string | null
+  activateTab: (id: string) => void
+  focusTab: (id: string) => void
+  getTabListProps: (props?: TabListProps) => HTMLAttributes<HTMLElement>
+  getTabProps: (id: string) => HTMLAttributes<HTMLElement>
+  getTabPanelProps: (id: string) => HTMLAttributes<HTMLElement>
+}
+
+export const useTabs = ({
+  defaultTab,
+  autoActivate = true,
+  orientation = 'horizontal',
+  onTabActivate,
+}: UseTabsOptions = {}): UseTabs => {
+  const uniqueId = useId()
+  const [state, _setState] = useState<TabState>({
+    activeTab: defaultTab || null,
+    focusedTab: null,
+    tabs: new Set(),
+  })
+
+  const setState = useCallback(
```

Comment:
Could we use something besides `setState` for this? That feels more a defacto reserved term and is needing you to prepend the usual setState with `_`. 

Perhaps call this `setTabState` instead?

Reply:
Happy to change it, but the choice of `setState` was intentional here as I want to encourage use of my custom `setState` wrapper and discourage use of the original `_setState` since using the raw `_setState` will prevent `onTabActivate` from being called

Reply:
Just checking that I've not misunderstood the comment. I'd be happy to change to the following if this is what you were suggesting:

```tsx
  const [tabState, _setTabState] = useState<TabState>({
    activeTab: defaultTab || null,
    focusedTab: null,
    tabs: new Set(),
  })

  const setTabState = useCallback(
```

Reply:
Yeah that works, thanks 👍

Reply:
Cool, updated c99b1d4a35daa8fa3448053fc0531a9424549899